From e3ef8ba802310f9a812183c575cf1b2c2caef790 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 22 Jun 2011 05:08:28 -0600
Subject: [PATCH 148/158] - fixed assert in cob_ns_lookup() to take into
 account hardlinks with zero nlink; - in c2_md_create() even if we found
 existing name belonging to our object we better   kill it and recreate if
 version allows to do so. It may be zombie and we need to   take this into
 account; - fixes in unlink. Attrs should be updated for killing hardlink.
 Before doing so   we check for connect nlink in attrs; - cleanups, comments
 and code simolificatons; - added assert in c2_md_store_create() to catch
 creating objects with zero nlink.

---
 cob/cob.c           |   3 +-
 mdservice/md_foms.c | 192 ++++++++++++++++++++++------------------------------
 mdstore/mdstore.c   |   1 +
 3 files changed, 83 insertions(+), 113 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index b1fd03b..6ac3283 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -405,7 +405,8 @@ static int cob_ns_lookup(struct c2_cob *cob, struct c2_db_tx *tx)
         
         if (rc == 0) {
                 cob->co_valid |= CA_NSREC;
-                C2_ASSERT(cob->co_nsrec.cnr_nlink > 0);
+                C2_ASSERT(cob->co_nsrec.cnr_linkno > 0 ||
+                          cob->co_nsrec.cnr_nlink > 0);
                 C2_POST(c2_fid_is_set(cob->co_fid));
         }
         return rc;
diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index 43ae95e..b013c92 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -139,6 +139,11 @@ static void c2_md_fop_attr2cob(struct c2_fop_cob *body,
         body->b_version = attr->ca_version;
 }
 
+/**
+   Create object in @pfid with @tfid and attributes from @attr.
+   Handle possible variants for existing/missing objects and
+   links.
+ */
 static int c2_md_create(struct c2_md_store  *md, 
                         struct c2_fid       *pfid,
                         struct c2_fid       *tfid, 
@@ -190,29 +195,22 @@ restart:
                  * Do we need to do anything at all?
                  */
                 if (attr->ca_version > scob->co_nsrec.cnr_version) {
-                        if (!c2_fid_eq(ncob->co_fid, tfid)) {
-                                /*
-                                 * Different object, this may be hardlink or
-                                 * result of racer work.
-                                 */
-                                rc = c2_md_store_unlink(md, pfid, scob, 
-                                                        attr->ca_name,
-                                                        attr->ca_namelen,
-                                                        tx);
-                                c2_cob_put(scob);
-                                c2_cob_put(ncob);
-                                if (rc)
-                                        return rc;
-                                goto restart;
-                        } else {
-                                /*
-                                 * Name that belongs to our object exists, it
-                                 * should be stat data or db is not consistent.
-                                 * We don't put an assert here because -EDEADLK
-                                 * may occur.
-                                 */
-                                rc = c2_md_store_setattr(md, scob, attr, tx);
-                        }
+                        /*
+                         * Name exists and does not matter if it
+                         * belongs to our object or not. If it is
+                         * ours then it should have to be marked
+                         * zombie. Anyways we want to kill it and
+                         * create again.
+                         */
+                        rc = c2_md_store_unlink(md, pfid, scob, 
+                                                attr->ca_name,
+                                                attr->ca_namelen,
+                                                tx);
+                        c2_cob_put(scob);
+                        c2_cob_put(ncob);
+                        if (rc)
+                                return rc;
+                        goto restart;
                 }
                 c2_cob_put(ncob);
         }
@@ -377,6 +375,8 @@ static int c2_md_unlink_fom_state(struct c2_fom *fom)
 
         body = &req->u_body;
         c2_md_fop_cob2attr(&attr, body);
+        attr.ca_name = req->u_name.s_buf;
+        attr.ca_namelen = req->u_name.s_len;
 
         c2_md_make_fid(&pfid, &body->b_pfid);
         c2_md_make_fid(&tfid, &body->b_tfid);
@@ -385,11 +385,11 @@ static int c2_md_unlink_fom_state(struct c2_fom *fom)
          * Mark name zombie, replicator will kill it later.
          */
         if (body->b_bias & C2_MD_BIAS_SCAN_IN_PROGRESS) {
-                rc = c2_md_store_lookup(md, &pfid, req->u_name.s_buf,
-                                        req->u_name.s_len, &ncob, tx);
+                rc = c2_md_store_lookup(md, &pfid, attr.ca_name,
+                                        attr.ca_namelen, &ncob, tx);
                 /*
                  * If no name found we need to create it with zombie
-                 * flag set to 1 or initial scan can find it later,
+                 * flag set to 1 or initial scan can create it later,
                  * which is not what we want.
                  */
                 if (rc == -ENOENT) {
@@ -411,37 +411,42 @@ static int c2_md_unlink_fom_state(struct c2_fom *fom)
                                  */
                                 if (rc == 0 &&
                                     attr.ca_version > scob->co_nsrec.cnr_version)
-                                        rc = c2_md_store_setattr(md, scob, &attr, tx);
+                                        rc = c2_md_store_setattr(md, scob, &attr,
+                                                                 tx);
                         }
                         if (scob)
                                 c2_cob_put(scob);
                 } else if (rc == 0) {
                         /*
-                         * Mark name zombie.
+                         * Mark name zombie, it will be killed after the scan.
                          */
                         ncob->co_nsrec.cnr_zombie = 1;
-                        rc = c2_cob_update(ncob, &ncob->co_nsrec, NULL, NULL, tx);
+                        rc = c2_cob_update(ncob, &ncob->co_nsrec, NULL, 
+                                           NULL, tx);
                         if (rc) {
                                 c2_cob_put(ncob);
                                 goto out;
                         }
                         /*
-                         * Update attributes for the case object will survive
-                         * (unlink of a hardlink). Name may be occupid by some
-                         * another object, let's use fid from ncob instead of
-                         * passed tfid.
+                         * Update attributes. Name may be occupid by some
+                         * another object, let's use fid from ncob instead
+                         * of passed tfid.
                          */
-                        if (ncob->co_nsrec.cnr_linkno > 0) {
-                                rc = c2_md_store_locate(md, ncob->co_fid, &scob, 
-                                                        C2_MD_LOCATE_STORED, tx);
-                                if (rc) {
-                                        c2_cob_put(ncob);
-                                        goto out;
-                                }
-                                if (attr.ca_version > scob->co_nsrec.cnr_version)
-                                        rc = c2_md_store_setattr(md, scob, &attr, tx);
-                                c2_cob_put(scob);
+                        rc = c2_md_store_locate(md, ncob->co_fid, &scob, 
+                                                C2_MD_LOCATE_STORED, tx);
+                        if (rc) {
+                                c2_cob_put(ncob);
+                                goto out;
                         }
+
+                        /*
+                         * We cannot update foreign object with our attributes.
+                         * Let's check this.
+                         */
+                        if (c2_fid_eq(&tfid, ncob->co_fid) && attr.ca_nlink > 0 &&
+                            attr.ca_version > scob->co_nsrec.cnr_version)
+                                rc = c2_md_store_setattr(md, scob, &attr, tx);
+                        c2_cob_put(scob);
                         c2_cob_put(ncob);
                 }
         } else {
@@ -451,10 +456,8 @@ static int c2_md_unlink_fom_state(struct c2_fom *fom)
                 if (rc)
                         goto out;
                 
-                rc = c2_md_store_unlink(md, &pfid, scob, 
-                                        req->u_name.s_buf,
-                                        req->u_name.s_len,
-                                        tx);
+                rc = c2_md_store_unlink(md, &pfid, scob, attr.ca_name,
+                                        attr.ca_namelen, tx);
                 if (rc == 0 && scob->co_nsrec.cnr_nlink > 0 &&
                     attr.ca_version > scob->co_nsrec.cnr_version)
                         rc = c2_md_store_setattr(md, scob, &attr, tx);
@@ -531,31 +534,24 @@ static int c2_md_rename_src(struct c2_md_store *md,
         rc = c2_md_store_lookup(md, pfid_src, sattr->ca_name, 
                                 sattr->ca_namelen, &sncob, tx);
         if (c2_fid_eq(tfid_tgt, tfid_src)) {
-                if (rc == -ENOENT) {
-                        if (scan) {
-                                /*
-                                 * Just created tgt with correct name. Let's
-                                 * make sure that scan will not re-create src.
-                                 */
-                                rc = c2_md_store_link(md, pfid_src, tcob,
-                                                      sattr->ca_name,
-                                                      sattr->ca_namelen,
-                                                      1, tx);
-                        }
+                if (rc == -ENOENT && scan) {
+                        /*
+                         * Just created tgt with correct name. Let's
+                         * make sure that scan will not re-create src.
+                         */
+                        rc = c2_md_store_link(md, pfid_src, tcob,
+                                              sattr->ca_name,
+                                              sattr->ca_namelen,
+                                              1, tx);
                 } else if (rc == 0) {
                         /*
-                         * Name exists, let's do normal rename and create old
-                         * name with zombie flag.
+                         * Name exists, let's do normal rename and create
+                         * old name with zombie flag.
                          */
                         rc = c2_md_rename(md, pfid_tgt, pfid_src, tfid_tgt,
                                           tfid_src, tattr, sattr, tcob, tcob,
                                           tx);
-                        if (rc) {
-                                c2_cob_put(sncob);
-                                return rc;
-                        }
-
-                        if (scan) {
+                        if (rc == 0 && scan) {
                                 /*
                                  * Src name is gone now but we want to make
                                  * sure that scan will not re-create it so
@@ -590,12 +586,7 @@ static int c2_md_rename_src(struct c2_md_store *md,
                                                   tfid_tgt, tfid_src,
                                                   tattr, sattr, tcob,
                                                   scob, tx);
-                                if (rc) {
-                                        c2_cob_put(scob);
-                                        return rc;
-                                }
-                                                
-                                if (scan) {
+                                if (rc == 0 && scan) {
                                         /*
                                          * Src name is gone now but we want to
                                          * make sure that scan will not re-create
@@ -629,12 +620,7 @@ static int c2_md_rename_src(struct c2_md_store *md,
                                                   tfid_tgt, tfid_src,
                                                   tattr, sattr, tcob,
                                                   scob, tx);
-                                if (rc) {
-                                        c2_cob_put(scob);
-                                        return rc;
-                                }
-                                                
-                                if (scan) {
+                                if (rc == 0 && scan) {
                                         /*
                                          * Src name is gone now but we want to
                                          * make sure that scan will not re-create
@@ -667,12 +653,7 @@ static int c2_md_rename_src(struct c2_md_store *md,
                                                   tfid_tgt, tfid_src,
                                                   tattr, sattr, tcob,
                                                   scob, tx);
-                                if (rc) {
-                                        c2_cob_put(scob);
-                                        return rc;
-                                }
-                                                
-                                if (scan) {
+                                if (rc == 0 && scan) {
                                         /*
                                          * Src name is gone now but we want to
                                          * make sure that scan will not re-create
@@ -693,18 +674,13 @@ static int c2_md_rename_src(struct c2_md_store *md,
                                                   tfid_tgt, tfid_src,
                                                   tattr, sattr, tcob,
                                                   scob, tx);
-                                if (rc) {
-                                        c2_cob_put(sncob);
-                                        c2_cob_put(scob);
-                                        return rc;
-                                }
                         }
                         c2_cob_put(sncob);
                 }
         }
         if (scob) {
-                if (sattr->ca_version > scob->co_nsrec.cnr_version &&
-                    scob->co_nsrec.cnr_nlink > 0)
+                if (rc == 0 && scob->co_nsrec.cnr_nlink > 0 && 
+                    sattr->ca_version > scob->co_nsrec.cnr_version)
                         rc = c2_md_store_setattr(md, scob, sattr, tx);
                 c2_cob_put(scob);
         }
@@ -790,27 +766,19 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
                                               &tattr, &sattr, tcob, 
                                               scan, tx);
         } else if (rc == 0) {
-                if (c2_fid_eq(tncob->co_fid, &tfid_tgt)) {
-                        rc = c2_md_store_locate(md, &tfid_tgt, &tcob, 
-                                                C2_MD_LOCATE_STORED, tx);
-                        if (rc == -ENOENT) {
-                                /*
-                                 * Target seems to be valid just missing
-                                 * (scan did not find it yet?). Let's create
-                                 * it.
-                                 */
-                                C2_ASSERT(tattr.ca_nlink > 0);
-                                rc = c2_md_store_create(md, &pfid_tgt, &tattr, 
-                                                        &tcob, tx);
-                        }
-                } else {
-                        /*
-                         * Target was just killed but there is some name.
-                         * Let's find it stat data and use as tcob.
-                         */
-                        rc = c2_md_store_locate(md, tncob->co_fid, &tcob, 
-                                                C2_MD_LOCATE_STORED, tx);
-                }
+                /*
+                 * Let's grab stat data from target name and use it 
+                 * for rename.
+                 */
+                rc = c2_md_store_locate(md, tncob->co_fid, &tcob, 
+                                        C2_MD_LOCATE_STORED, tx);
+
+                /*
+                 * Create stat data name if it is not existing and name
+                 * we've found belongs to our object.
+                 */
+                if (rc == -ENOENT && c2_fid_eq(tncob->co_fid, &tfid_tgt))
+                        rc = c2_md_store_create(md, &pfid_tgt, &tattr, &tcob, tx);
 
                 if (rc == 0)
                         rc = c2_md_rename_src(md, &pfid_tgt, &pfid_src, 
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 0559d1e..2001673 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -137,6 +137,7 @@ int c2_md_store_create(struct c2_md_store       *md,
                           attr->ca_namelen);
 
         nsrec.cnr_fid = attr->ca_tfid;
+        C2_ASSERT(attr->ca_nlink > 0);
         nsrec.cnr_nlink = attr->ca_nlink;
         nsrec.cnr_rdev = attr->ca_rdev;
         nsrec.cnr_size = attr->ca_size;
-- 
1.8.3.2

